File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/majority_element.py
"""Majority element by recursive candidate comparison."""


def find(values: list[int]) -> int | None:
    """Return the value appearing more than n/2 times, or None."""
    if not values:
        raise ValueError("values must not be empty")
    candidate = _candidate(values, 0, len(values) - 1)
    return candidate if values.count(candidate) > len(values) // 2 else None


def _candidate(values: list[int], left: int, right: int) -> int:
    if left == right:
        return values[left]
    mid = left + (right - left) // 2
    left_candidate = _candidate(values, left, mid)
    right_candidate = _candidate(values, mid + 1, right)
    if left_candidate == right_candidate:
        return left_candidate
    left_count = sum(1 for index in range(left, right + 1) if values[index] == left_candidate)
    right_count = sum(1 for index in range(left, right + 1) if values[index] == right_candidate)
    return left_candidate if left_count >= right_count else right_candidate

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/MajorityElement.java
package com.divideconquer.algorithms;

import java.util.OptionalInt;

public final class MajorityElement {
    private MajorityElement() {
    }

    public static OptionalInt find(int[] values) {
        if (values == null || values.length == 0) {
            throw new IllegalArgumentException("values must not be empty");
        }
        int candidate = majorityCandidate(values, 0, values.length - 1);
        int count = 0;
        for (int value : values) {
            if (value == candidate) {
                count++;
            }
        }
        return count > values.length / 2 ? OptionalInt.of(candidate) : OptionalInt.empty();
    }

    private static int majorityCandidate(int[] values, int left, int right) {
        if (left == right) {
            return values[left];
        }
        int mid = left + (right - left) / 2;
        int leftCandidate = majorityCandidate(values, left, mid);
        int rightCandidate = majorityCandidate(values, mid + 1, right);
        if (leftCandidate == rightCandidate) {
            return leftCandidate;
        }
        int leftCount = countInRange(values, leftCandidate, left, right);
        int rightCount = countInRange(values, rightCandidate, left, right);
        return leftCount >= rightCount ? leftCandidate : rightCandidate;
    }

    private static int countInRange(int[] values, int target, int left, int right) {
        int count = 0;
        for (int i = left; i <= right; i++) {
            if (values[i] == target) {
                count++;
            }
        }
        return count;
    }
}
